driver.go

```diff
@@ -64,94 +62,230 @@ func (d *Driver) Close() {
+		idrec, err := tx.Run(query, vars)
 
-func buildPropertyMap(in map[string]interface{}, out map[string]*ls.PropertyValue) {
-	for k, v := range in {
-		switch k {
-		case PropNodeValue:
-		case PropNodeID:
-		default:
-			switch val := v.(type) {
-			case []interface{}:
-				arr := make([]string, 0, len(val))
-				for _, x := range val {
-					arr = append(arr, fmt.Sprint(x))
-				}
-				out[k] = ls.StringSlicePropertyValue(arr)
-			default:
-				out[k] = ls.StringPropertyValue(fmt.Sprint(val))
-			}
+		if err != nil {
+			return err
+		}
+		rec, err := idrec.Single()
+		if err != nil {
+			return err
+		}
+		nd := rec.Values[0].(neo4j.Node)
+		nodeIds[edge.GetTo()] = nd.Id
+		return nil
+	}
+	// contains only target node
+	if !contains(edge.GetFrom(), nodeIds) && contains(edge.GetTo(), nodeIds) {
+		// (newNode) --edge-->(match) --edge-->(newNode)
+		vars := make(map[string]interface{})
+		query := fmt.Sprintf("MATCH (to) WHERE ID(to) = %d CREATE (to)<-[%s %s]-(from %s %s) RETURN from",
+			nodeIds[edge.GetTo()], makeLabels(vars, []string{edge.GetLabel()}), makeProperties(vars, ls.PropertiesAsMap(edge), nil),
+			makeLabels(vars, edge.GetFrom().GetLabels().Slice()), makeProperties(vars, ls.PropertiesAsMap(edge.GetFrom()), nil))
+		idrec, err := tx.Run(query, vars)
+		if err != nil {
+			return err
+		}
+		rec, err := idrec.Single()
+		if err != nil {
+			return err
 		}
+		nd := rec.Values[0].(neo4j.Node)
+		nodeIds[edge.GetFrom()] = nd.Id
+		return nil
 	}
+	// source,target does not exist in db
+	// (newNode) --edge-->(newNode)
+	fromId, toId, err := s.CreateNodePair(tx, edge)
+	if err != nil {
+		return err
+	}
+	nodeIds[edge.GetFrom()] = fromId
+	nodeIds[edge.GetTo()] = toId
+	return nil
 }
 
-// CreateNode creates a new node and returns its neo4j ID
-func (s *Session) CreateNode(tx neo4j.Transaction, node ls.Node) (int64, error) {
-	vars := make(map[string]interface{})
-	labelsClause := makeLabels(node.GetTypes().Slice())
-	idAndValue := make(map[string]*ls.PropertyValue)
-	if len(node.GetID()) > 0 {
-		idAndValue[PropNodeID] = ls.StringPropertyValue(node.GetID())
+func (s *Session) emptyDB(tx neo4j.Transaction) bool {
+	idrec, err := tx.Run("MATCH (n) RETURN count(n)", make(map[string]interface{}))
+	if err != nil {
+		return false
 	}
-	if node.GetValue() != nil {
-		idAndValue[PropNodeValue] = ls.StringPropertyValue(fmt.Sprint(node.GetValue()))
+	rec, err := idrec.Single()
+	if err != nil {
+		return false
 	}
-	propertiesClause := makeProperties(vars, node.GetProperties(), idAndValue)
-	query := fmt.Sprintf("CREATE (n %s %s) RETURN n", labelsClause, propertiesClause)
-	s.Logf("Create node %s with %v", query, vars)
+	nd := rec.Values[0].(int64)
+	if nd == 0 {
+		return true
+	}
+	return false
+}
+
+func (s *Session) existsDB(tx neo4j.Transaction, node graph.Node) (int64, error) {
+	if node == nil {
+		return 0, nil
+	}
+	vars := make(map[string]interface{})
```

Comment:
This should lookup a node using entity id and labels. So the query should look like:

```
match (n labels props) return n
```
where:
   *  labels is all labels of the node,
   *  props is {https://lschema.org/entityId: $entityIdValue}, where entityIdValue is either the string property value, or []string property value of the entityId term in the node,